contrib/babelfishpg_tsql/src/fts_parser.y

```diff
@@ -113,71 +117,217 @@ simple_term_list:
+                                                ereport(ERROR,
+                                                        (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+                                                         errmsg("Concurrent special characters in the full-text search condition are not currently supported in Babelfish")));
+                                        } 
+                                }
+                                appendStringInfoChar(&modifiedInput, input[i]);
+                        }
+                }
+                /* Null-terminate the modified input string */
+                appendStringInfoChar(&modifiedInput, '\0');
+        } else {
+                appendStringInfoString(&modifiedInput, input);
         }
-        // It's a single word, so no transformation needed
-        return trimmedInputStr;
-    }
+
+        inputLen = strlen(modifiedInput.data);
+
+        for (i = 0; i < inputLen; i++) {
+                char currentChar = modifiedInput.data[i];
+
+                /* Store the result of strchr to avoid calling it multiple times */
+                charInForbiddenChars = strchr(forbiddenChars, currentChar);
+                charInSpecialChars = strchr(specialChars, currentChar);
+                charInBoolOperators = strchr(boolOperators, currentChar);
+
+                if (currentChar == '~') {
+                        ereport(ERROR,
+                                (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+                                 errmsg("Proximity term is not currently supported in Babelfish")));
+                }
+
+                if (charInBoolOperators != NULL) {
+                        ereport(ERROR,
+                                (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+                                 errmsg("Full-text search conditions with boolean operators are not currently supported in Babelfish")));
+                }
+
+                /* Check for forbidden characters when not in quotes */
+                if (!isEnclosedInQuotes && charInForbiddenChars != NULL) {
+                        ereport(ERROR,
+                                (errcode(ERRCODE_SYNTAX_ERROR),
+                                 errmsg("Syntax error near '%c' in the full-text search condition '%s'.", currentChar, input)));
+                }
+
+                /* Check for consecutive special characters */
+                if (charInSpecialChars != NULL && i + 1 < inputLen && strchr(specialChars, modifiedInput.data[i + 1]) != NULL) {
+                        ereport(ERROR,
+                                (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+                                 errmsg("Concurrent special characters in the full-text search condition are not currently supported in Babelfish")));
+                } else if (isspace(currentChar) || charInSpecialChars != NULL || (isEnclosedInQuotes && strchr("`'_", currentChar) != NULL)) {
+                        if (!inSpace) {
```

Comment:
where are we detecting concurrent special characters when space is in-between?

Reply:
We are detecting it in the if-clause that is above this for loop. This one:
```
if (strchr(specialChars, input[i]) != NULL) {
        while (i + 1 < inputLen && isspace(input[i + 1])) {
                i++;
        }
        if (i + 1 < inputLen && strchr(specialChars, input[i + 1]) != NULL) {
                ereport(ERROR,
                        (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
                         errmsg("Concurrent special characters in the full-text search condition are not currently supported in Babelfish")));
        } 
}
appendStringInfoChar(&modifiedInput, input[i]);
```